### PR TITLE
feat(pd): add pre-prefill routing for PD disaggregation

### PR DESCRIPTION
**Pull request by kzjeef**
## Summary

Add pre-prefill routing to optimize long-context request handling in PD disaggregation.

- Cold requests (low cache match ratio) are routed to a dedicated pre-prefill worker that warms up KV cache asynchronously
- Future prefill requests with overlapping prefixes hit the warmed cache, reducing TTFT
- Routing decision based on 3 configurable thresholds: `match_threshold`, `unmatched_chars_threshold`, `min_tokens`

## Background

In PD disaggregation, long-context requests that have no KV cache locality (cold requests) cause high TTFT because the prefill worker must process the full context from scratch. Pre-prefill routing solves this by:

1. Detecting cold requests via the cache-aware radix tree (low match ratio + enough unmatched content)
2. Routing them to a dedicated **pre-prefill worker** that asynchronously warms the L3 KV cache
3. Recording the assignment in the radix tree so future requests with overlapping prefixes are recognized as warm
4. Warm requests bypass pre-prefill and go through normal cache-aware routing to the worker that already has the relevant cache

This is particularly useful for RAG/long-document workloads where the first request for a new document is expensive, but subsequent queries against the same document should be fast.

## Architecture

```
                    ┌─────────────────────────────┐
                    │     Incoming Chat Request     │
                    └──────────────┬──────────────┘
                                   │
                    ┌──────────────▼──────────────┐
                    │  text.len() >= min_tokens?   │
                    └──────┬───────────────┬──────┘
                           │ NO            │ YES
                           │               ▼
                           │    ┌─────────────────────┐
                           │    │ estimate_match_stats │
                           │    │  (radix tree lookup) │
                           │    └────┬───────────┬────┘
                           │         │           │
                           │   match_ratio    match_ratio
                           │    >= 0.1         < 0.1 AND
                           │   (WARM)       unmatched >= 10K
                           │         │        (COLD)
                           │         │           │
                           ▼         ▼           ▼
                    ┌──────────┐ ┌────────┐ ┌──────────────┐
                    │  Normal  │ │ Normal │ │ Pre-Prefill  │
                    │ Cache-   │ │ Cache- │ │   Worker     │
                    │  Aware   │ │ Aware  │ │ (warm cache) │
                    │ Routing  │ │Routing │ │              │
                    └──────────┘ └────────┘ └──────────────┘
```

### New fields on `RoutingMode::PrefillDecode`

| Field | Type | Default | Description |
|-------|------|---------|-------------|
| `pre_prefill_url` | `Option<String>` | `None` | URL of the pre-prefill worker. Feature disabled when `None`. |
| `pre_prefill_decode_url` | `Option<String>` | `None` | Decode worker paired with the pre-prefill worker. |
| `pre_prefill_match_threshold` | `f32` | `0.1` | Cache match ratio below which a request is considered cold. |
| `pre_prefill_unmatched_chars_threshold` | `usize` | `10000` | Minimum unmatched characters to trigger pre-prefill. |
| `pre_prefill_min_tokens` | `usize` | `10000` | Minimum text length (chars) to even consider pre-prefill. Short requests always bypass. |

### Routing decision

A request is routed to the pre-prefill worker when ALL conditions are met:
1. `pre_prefill_url` is configured (feature enabled)
2. `text.len() >= pre_prefill_min_tokens` (request is long enough)
3. Prefill policy is `cache_aware` (supports radix tree lookup)
4. `match_ratio < pre_prefill_match_threshold` (low cache hit)
5. `unmatched_chars >= pre_prefill_unmatched_chars_threshold` (enough new content to justify warming)

Otherwise, normal cache-aware routing is used.

### CLI

```bash
# Basic PD mode with pre-prefill enabled
smg launch --pd-disaggregation \
  --prefill http://prefill1:8000 \
  --prefill http://prefill2:8000 \
  --decode http://decode1:8000 \
  --decode http://decode2:8000 \
  --pre-prefill-url http://pre-prefill:8000 \
  --pre-prefill-decode-url http://pre-prefill-decode:8000 \
  --policy cache_aware

# With custom thresholds (more aggressive cold detection)
smg launch --pd-disaggregation \
  --prefill http://prefill1:8000 \
  --decode http://decode1:8000 \
  --pre-prefill-url http://pre-prefill:8000 \
  --pre-prefill-decode-url http://pre-prefill-decode:8000 \
  --pre-prefill-match-threshold 0.2 \
  --pre-prefill-unmatched-chars-threshold 5000 \
  --pre-prefill-min-tokens 5000 \
  --policy cache_aware

# Without pre-prefill (existing behavior, no change)
smg launch --pd-disaggregation \
  --prefill http://prefill1:8000 \
  --decode http://decode1:8000 \
  --policy cache_aware
```

### Python

```python
from smg.smg_rs import Router, PolicyType

router = Router(
    worker_urls=[],
    policy=PolicyType.CacheAware,
    pd_disaggregation=True,
    prefill_urls=[
        ("http://prefill1:8000", None),
        ("http://prefill2:8000", None),
    ],
    decode_urls=[
        "http://decode1:8000",
        "http://decode2:8000",
    ],
    # Pre-prefill configuration
    pre_prefill_url="http://pre-prefill:8000",
    pre_prefill_decode_url="http://pre-prefill-decode:8000",
    pre_prefill_match_threshold=0.1,
    pre_prefill_unmatched_chars_threshold=10000,
    pre_prefill_min_tokens=10000,
)
router.start()
```

### Python CLI (via argparse)

```bash
python -m sglang_router.launch_router \
  --pd-disaggregation \
  --prefill http://prefill1:8000 \
  --decode http://decode1:8000 \
  --pre-prefill-url http://pre-prefill:8000 \
  --pre-prefill-decode-url http://pre-prefill-decode:8000 \
  --policy cache_aware
```

## Files changed (11)

| File | Change |
|------|--------|
| `config/types.rs` | 5 new fields on `PrefillDecode`, 3 default fns |
| `config/builder.rs` | Builder methods updated with defaults |
| `config/validation.rs` | Match arms use `..` for new fields |
| `main.rs` | 5 CLI args under "PD Disaggregation" heading |
| `policies/cache_aware.rs` | `estimate_match_stats()` + `record_assignment()` + 7 unit tests |
| `routers/http/pd_router.rs` | `extract_chat_request_text`, `select_prefill_worker`, `select_decode_worker` + 10 unit tests |
| `bindings/python/src/lib.rs` | Python bindings for 5 new fields |
| `bindings/python/src/smg/router_args.py` | Dataclass fields + argparse args |
| `tests/common/mock_worker.rs` | Per-port request counters for routing verification |
| `tests/routing/pd_routing_test.rs` | 2 integration tests with mock sglang workers |
| `tests/routing/test_pd_routing.rs` | Updated existing constructors |

## Test plan

- [x] **Unit tests** (17 new): `estimate_match_stats`, `record_assignment`, `extract_chat_request_text`, routing decision paths (cold, warm, short bypass, fallback, no text)
- [x] **Integration test** `test_pre_prefill_cold_routes_to_pp_worker`: verifies cold → PP worker via per-port request counters on mock sglang workers
- [x] **Integration test** `test_pre_prefill_multi_turn_conversation_simulation`: simulates 3 interleaved conversations (7 requests), verifies:
  - New conversations (cold) → PP worker (match_ratio: 0.000, 0.052, 0.060)
  - Follow-up turns (warm) → normal routing (match_ratio: 0.426 → 0.713 → 0.823)
- [x] All 841 existing tests pass, 0 failures
- [x] `cargo build` clean, no new warnings

### Integration test trace (multi-turn simulation)

```
Conv A T1 (cold)  match=0.000 → PP ✓   New conversation, no cache
Conv A T2 (warm)  match=0.426 → Normal ✓   Shares 42% prefix with T1
Conv B T1 (cold)  match=0.052 → PP ✓   New conversation
Conv A T3 (warm)  match=0.713 → Normal ✓   71% prefix overlap (3 turns)
Conv C T1 (cold)  match=0.060 → PP ✓   New conversation
Conv B T2 (warm)  match=0.476 → Normal ✓   48% prefix overlap
Conv A T4 (warm)  match=0.823 → Normal ✓   82% prefix overlap (4 turns)
```

## Summary by CodeRabbit

* **New Features**
  * Added pre-prefill routing with configurable URLs, decode endpoints, match thresholds and token limits
  * Intelligent routing that sends cold/warm requests to pre-prefill and paired decode workers based on request text and cache-match statistics
  * New worker types and runtime wiring so pre-prefill stages participate in readiness, load and routing

* **Chores**
  * CLI and Python interfaces extended with flags/parameters and sensible defaults; validation added

* **Tests**
  * Comprehensive unit and integration tests covering pre-prefill flows, routing behaviors and per-worker request counting

### CODE REVIEW

File: model_gateway/src/routers/http/pd_router.rs [hunk 1]
```diff
@@ -156,13 +167,53 @@ impl PDRouter {
...
+        {
+            (
+                pre_prefill_url.clone(),
+                pre_prefill_decode_url.clone(),
+                *pre_prefill_match_threshold,
+                *pre_prefill_unmatched_chars_threshold,
+                *pre_prefill_min_tokens,
+            )
+        } else {
+            (None, None, 0.1, 10000, 10000)
```
[the remark this answers is not in the document]
Comment (kzjeef) [reply model_gateway/src/routers/http/pd_router.rs]: Fixed. Changed to `(None, None, 0.0, 0, 0)` — these values are inert since `has_pre_prefill()` returns false when `pre_prefill_url` is `None`.

File: model_gateway/src/routers/http/pd_router.rs [hunk 2]
```diff
@@ -697,6 +748,181 @@ impl PDRouter {
...
+        let decode_policy = self.policy_registry.get_decode_policy();
+        Self::pick_worker_by_policy_arc(
+            decode_workers,
+            &*decode_policy,
+            request_text,
+            headers,
+            hash_ring,
+            "decode",
+        )
+    }
```
[the remark this answers is not in the document]
Comment (kzjeef) [reply model_gateway/src/routers/http/pd_router.rs:928]: Fixed. Removed the random fallback entirely. The `select_decode_worker` now always falls through to the normal `decode_policy` via `pick_worker_by_policy_arc` when the paired decode worker is unavailable. This also removes the `rand::Rng` import.

File: model_gateway/src/routers/http/pd_router.rs [hunk 3]
```diff
@@ -697,6 +748,181 @@ impl PDRouter {
...
+                        );
+
+                        // Route to pre-prefill if the request is "cold"
+                        if match_ratio < self.pre_prefill_match_threshold
+                            && unmatched >= self.pre_prefill_unmatched_chars_threshold
+                        {
+                            // Find the pre-prefill worker in the available workers
+                            if let Some(worker) = prefill_workers
+                                .iter()
+                                .find(|w| w.url() == pre_prefill_url.as_str() && w.is_available())
```
[the remark this answers is not in the document]
Comment (kzjeef) [reply model_gateway/src/routers/http/pd_router.rs]: Good catch. The pre-prefill URL must be included in `prefill_urls` (and the decode URL in `decode_urls`) for the worker to be registered and discoverable. I've added a startup warning in `PDRouter::new()` when `pre_prefill_decode_url` is missing. Adding a config-time validation that checks URL membership would require knowing the resolved worker list at validation time, which isn't available yet (workers are registered asynchronously via job queue). The current approach fails safely — if the URL isn't found in the worker list, routing falls back to normal policy.

File: model_gateway/src/config/validation.rs [hunk 4]
```diff
@@ -112,6 +112,7 @@ impl ConfigValidator {
                 decode_urls,
                 prefill_policy,
                 decode_policy,
+                ..
```
[the remark this answers is not in the document]
Comment (kzjeef) [reply model_gateway/src/config/validation.rs:150]: Fixed. Added range validation for `pre_prefill_match_threshold` (must be 0.0-1.0) in `validate_mode()`.

File: bindings/python/src/smg/router_args.py [hunk 5]
```diff
@@ -23,6 +23,12 @@ class RouterArgs:
         default_factory=list
     )  # List of (url, bootstrap_port)
     decode_urls: list[str] = dataclasses.field(default_factory=list)
+    # Pre-prefill routing configuration
+    pre_prefill_url: str | None = None  # URL for pre-prefill worker
+    pre_prefill_decode_url: str | None = None  # Decode URL paired with pre-prefill worker
+    pre_prefill_match_threshold: float = 0.1  # Cache match threshold for cold detection
+    pre_prefill_unmatched_chars_threshold: int = 10000  # Min unmatched chars to trigger
+    pre_prefill_min_tokens: int = 10000  # Min total tokens for eligibility
```
[the remark this answers is not in the document]
Comment (kzjeef) [reply bindings/python/src/smg/router_args.py]: Fixed. Added a startup warning in `PDRouter::new()` when `pre_prefill_url` is set but `pre_prefill_decode_url` is missing.

File: model_gateway/src/config/builder.rs [hunk 6]
```diff
@@ -60,6 +60,11 @@ impl RouterConfigBuilder {
             decode_urls,
             prefill_policy: None,
             decode_policy: None,
+            pre_prefill_url: None,
+            pre_prefill_decode_url: None,
+            pre_prefill_match_threshold: 0.1,
+            pre_prefill_unmatched_chars_threshold: 10000,
+            pre_prefill_min_tokens: 10000,
```
[the remark this answers is not in the document]
Comment (kzjeef) [reply model_gateway/src/config/builder.rs]: Fixed. Made `default_pre_prefill_*()` functions `pub(crate)` in types.rs and updated builder.rs to call them instead of hardcoding literals.

File: model_gateway/src/routers/http/pd_router.rs [hunk 7]
```diff
@@ -697,6 +748,181 @@ impl PDRouter {
...
+
+                            debug!(
+                                "Pre-prefill worker {} not available, falling back to normal routing",
+                                pre_prefill_url
+                            );
+                        }
+                    }
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (kzjeef) [reply model_gateway/src/routers/http/pd_router.rs]: Fixed. Added `warn\!()` in `PDRouter::new()` when pre-prefill is configured but prefill policy isn't `cache_aware`.

File: model_gateway/src/routers/http/pd_router.rs [hunk 8]
```diff
@@ -1279,19 +1526,8 @@ impl RouterTrait for PDRouter {
...
-                    MessageContent::Parts(_) => None,
-                },
-                ChatMessage::System { content, .. } => Some(content.to_simple_string()),
-                _ => None,
-            })
+        let request_text = if self.policies_need_request_text() || self.has_pre_prefill() {
+            Self::extract_chat_request_text(&body.messages)
         } else {
             None
         };
```
[the remark this answers is not in the document]
Comment (kzjeef) [reply model_gateway/src/routers/http/pd_router.rs]: Fixed. Added `|| self.has_pre_prefill()` to both `route_generate` and `route_completion` so pre-prefill text extraction works for all request types.

File: model_gateway/tests/routing/pd_routing_test.rs [hunk 9]
```diff
@@ -219,4 +220,406 @@ mod pd_routing_tests {
...
+    /// which worker actually received each request through the full HTTP stack.
+    ///
+    /// The cache-aware tree is lazily initialized: the first request creates
+    /// the tree entry. So we send a warmup request first to seed the tree,
+    /// then a cold request (different text) that triggers pre-prefill.
+    #[tokio::test]
+    async fn test_pre_prefill_cold_routes_to_pp_worker() {
+        mock_worker::reset_request_counters();
+
+        let prefill_port_1 = 19830;
```
[the remark this answers is not in the document]
Comment (kzjeef) [reply model_gateway/tests/routing/pd_routing_test.rs:238]: Fixed. Added `#[serial]` (from `serial_test` crate, already in Cargo.toml) to both counter-based integration tests to prevent global state races.

File: model_gateway/src/routers/http/pd_router.rs [hunk 10]
```diff
@@ -697,6 +748,181 @@ impl PDRouter {
...
+                }
+                _ => {}
+            }
+        }
+        if result.is_empty() {
+            None
+        } else {
+            Some(result)
+        }
+    }
```
[the remark this answers is not in the document]
Comment (kzjeef) [reply model_gateway/src/routers/http/pd_router.rs:813]: Intentional. The radix tree does character-level prefix matching, so separators would reduce match ratios for multi-turn conversations. No separators gives the best prefix overlap between turns.

File: model_gateway/src/routers/http/pd_router.rs [hunk 11]
```diff
@@ -697,6 +748,181 @@ impl PDRouter {
...
+                                    pre_prefill_url, match_ratio, unmatched
+                                );
+                                // Record the assignment so future similar requests hit cache
+                                cache_aware.record_assignment(
+                                    prefill_workers,
+                                    text,
+                                    worker.url(),
+                                );
+                                return Ok((worker.clone(), true));
+                            }
```
[the remark this answers is not in the document]
Comment (kzjeef) [reply model_gateway/src/routers/http/pd_router.rs]: Agree it could be noisy. Keeping it at `info\!` for now since cold requests should be relatively rare in steady state. If it becomes an issue in production we can downgrade to `debug\!`.

File: model_gateway/src/routers/http/pd_router.rs [hunk 12]
```diff
@@ -697,6 +748,181 @@ impl PDRouter {
...
+                .iter()
+                .filter(|w| w.is_available())
+                .collect();
+            if available.is_empty() {
+                return Err("No available decode workers".to_string());
+            }
+            let mut rng = rand::rng();
+            let idx = rng.random_range(0..available.len());
+            return Ok(available[idx].clone());
+        }
```
[the remark this answers is not in the document]
Comment (kzjeef) [reply model_gateway/src/routers/http/pd_router.rs]: Fixed. The random fallback has been removed entirely — decode selection now always goes through the configured decode policy.

File: model_gateway/tests/routing/pd_routing_test.rs [hunk 13]
```diff
@@ -219,4 +220,406 @@ mod pd_routing_tests {
...
+            .host("127.0.0.1")
+            .port(3803)
+            .max_payload_size(256 * 1024 * 1024)
+            .request_timeout_secs(600)
+            .worker_startup_timeout_secs(5)
+            .worker_startup_check_interval_secs(1)
+            .max_concurrent_requests(64)
+            .queue_timeout_secs(60)
+            .build_unchecked();
+
```
[the remark this answers is not in the document]
Comment (kzjeef) [reply model_gateway/tests/routing/pd_routing_test.rs:271]: Fixed via `#[serial]` on both counter-based tests. The port numbers are also unique per test to avoid bind conflicts.

File: model_gateway/src/routers/http/pd_router.rs [hunk 14]
```diff
@@ -697,6 +763,169 @@ impl PDRouter {
...
+        let mut result = String::new();
+        for msg in messages {
+            match msg {
+                ChatMessage::User { content, .. }
+                | ChatMessage::System { content, .. }
+                | ChatMessage::Developer { content, .. } => {
+                    let text = content.to_simple_string();
+                    if !text.is_empty() {
+                        result.push_str(&text);
+                    }
```
[the remark this answers is not in the document]
Comment (kzjeef) [reply model_gateway/src/routers/http/pd_router.rs]: Intentional — no delimiters gives the best prefix overlap for multi-turn cache matching. The radix tree matches character-by-character, so adding delimiters would reduce match ratios when comparing turn N vs turn N+1.

File: model_gateway/src/routers/http/pd_router.rs [hunk 15]
```diff
@@ -697,6 +763,169 @@ impl PDRouter {
...
+                ChatMessage::Assistant { content, .. } => {
+                    if let Some(content) = content {
+                        let text = content.to_simple_string();
+                        if !text.is_empty() {
+                            result.push_str(&text);
+                        }
+                    }
+                }
+                _ => {}
+            }
```
[the remark this answers is not in the document]
Comment (kzjeef) [reply model_gateway/src/routers/http/pd_router.rs]: Tool/function content is excluded intentionally. Tool call arguments and outputs are typically short and variable (JSON payloads, API responses), which would add noise to the prefix matching without improving cache locality. The routing decision should be based on the human-readable conversation context.

File: model_gateway/src/routers/http/pd_router.rs [hunk 16]
```diff
@@ -50,6 +50,16 @@ pub struct PDRouter {
     pub retry_config: RetryConfig,
     pub api_key: Option<String>,
     pub enable_igw: bool,
+    /// URL for the pre-prefill worker (cold request warming)
+    pub pre_prefill_url: Option<String>,
```
Comment (slin1237) [inline model_gateway/src/routers/http/pd_router.rs]: we have the notion of worker
and worker can be different types, we currently have prefill, decode, regular
if we intended to add 2 more (p-prefill, p-decode)
just add them
and they can registered in worker registry
rather than adding urls here

**Review by slin1237 — CHANGES_REQUESTED** [review]

**Comment by slin1237** [comment]
The cache-aware policy additions (`estimate_match_stats`, `record_assignment`) are solid — they follow existing patterns, handle edge cases, and the mesh sync is consistent. The routing logic itself is sound. Tests are thorough, especially the multi-turn simulation. Issues below.

---

### 1. Hardcoded URLs bypass the worker abstraction

We have `WorkerType` (Prefill, Decode, Regular) and `WorkerRegistry` for type-based worker lookup. This PR adds `pre_prefill_url` / `pre_prefill_decode_url` as raw strings and locates workers by URL matching:

```rust
prefill_workers.iter().find(|w| w.url() == pre_prefill_url.as_str())
```

Pre-prefill should be new `WorkerType` variants (`PrePrefill`, `PrePrefillDecode`), registered in `WorkerRegistry` and queried via `get_by_type()` like everything else. That eliminates the 5 new fields on `RoutingMode::PrefillDecode`, the 5 CLI args, the 5 Python binding params, and all the boilerplate across ~15 test constructors.

### 2. Breaking change to text extraction for all cache-aware users

This is the most dangerous issue. The old ChatCompletions handler extracts text from the **first message only**:

```rust
// OLD
body.messages.first().and_then(|msg| match msg { ... })
```

The new code concatenates **ALL messages**:

```rust
// NEW
Self::extract_chat_request_text(&body.messages)
```

This applies whenever `policies_need_request_text()` is true — which it already is for `CacheAwarePolicy`. So every existing PD deployment using cache-aware routing (even without pre-prefill configured) now gets different text fed into the radix tree. Prefix matching characteristics change completely:

- Before: matching against `"You are a helpful assistant"` (short, stable)
- After: matching against `"You are a helpful assistantHello how are youI'm good..."` (grows every turn)

This silently alters cache-aware routing behavior for all users. The text extraction change needs to be gated behind `has_pre_prefill()` — keep the old first-message extraction for the normal policy path.

### 3. `pre_prefill_min_tokens` is compared against characters

```rust
if text.len() >= self.pre_prefill_min_tokens {
```

`text.len()` returns bytes/chars, not tokens. Either rename to `pre_prefill_min_chars` or actually tokenize. Current name will mislead operators tuning thresholds.

### 4. Threshold config in the wrong place

`pre_prefill_match_threshold`, `pre_prefill_unmatched_chars_threshold`, and `pre_prefill_min_tokens` are cache-aware policy tuning knobs. They belong in `PolicyConfig::CacheAware` or a dedicated `PrePrefillConfig`, not on `RoutingMode::PrefillDecode`. Routing mode describes topology (which workers), not policy behavior.

### 5. Nesting depth in `select_prefill_worker`

6 levels deep:

```rust
if let (Some(url), Some(text)) = ... {
    if text.len() >= ... {
        if let Some(cache_aware) = ... {
            if let Some((matched, total)) = ... {
                if match_ratio < ... && unmatched >= ... {
                    if let Some(worker) = ... {
```

Flatten with early returns — that's the Rust idiom.

### 6. Code duplication in `select_pd_pair`

The `has_pre_prefill()` branching duplicates the original `pick_worker_by_policy_arc` calls in the else branches. `select_prefill_worker` already falls back to normal selection internally, so always call it and let it handle the no-pre-prefill case.

### 7. Minor: no separator in text concatenation

`extract_chat_request_text` concatenates messages without any delimiter — `"Hello" + "World"` → `"HelloWorld"`. Creates phantom prefixes across message boundaries. A `\n` separator would be trivial and correct.

### 8. Minor: memory allocation on every request

When pre-prefill is enabled, every chat request concatenates ALL messages into one string. For a 20-turn conversation that's potentially 50-100KB allocated per request just for the cold check. If the request is warm, the allocation is wasted. Consider checking the first message as a fast path before doing the full concatenation.

File: model_gateway/src/routers/http/pd_router.rs (hunk 16 above)
Comment (kzjeef) [reply model_gateway/src/routers/http/pd_router.rs]: Good point. The current approach stores pre-prefill URLs as string fields to minimize the change surface — adding new WorkerType variants would touch WorkerRegistry, service discovery, health checks, metrics labels, job queue, and all the match arms across the codebase.

I'll update this PR later.
